src/qt/intro.cpp

```diff
@@ -0,0 +1,324 @@
+            } else {
+                replyStatus = ST_ERROR;
+                replyMessage = tr("Path already exists, and is not a directory.");
+            }
+        }
+    } catch (const fs::filesystem_error&)
+    {
+        /* Parent directory does not exist or is not accessible */
+        replyStatus = ST_ERROR;
+        replyMessage = tr("Cannot create data directory here.");
+    }
+    Q_EMIT reply(replyStatus, replyMessage, freeBytesAvailable);
+}
+
+Intro::Intro(QWidget *parent) :
+    QDialog(parent),
+    ui(new Ui::Intro),
+    thread(nullptr),
+    signalled(false)
+{
+
+    ui->setupUi(this);
+
+    ui->welcomeLabel->setText(ui->welcomeLabel->text().arg(PACKAGE_NAME));
+    ui->storageLabel->setText(ui->storageLabel->text().arg(PACKAGE_NAME));
+
+    ui->lblExplanation1->setText(ui->lblExplanation1->text().arg(PACKAGE_NAME));
+    ui->lblExplanation2->setText(ui->lblExplanation2->text().arg(PACKAGE_NAME));
+
+    // This is necessary to deal with various dpi displays.
+    this->adjustSize();
+
+    startThread();
+}
+
+Intro::~Intro()
+{
+    delete ui;
+    /* Ensure thread is finished before it is deleted */
+    thread->quit();
+    thread->wait();
+}
+
+QString Intro::getDataDirectory()
+{
+    return ui->dataDirectory->text();
+}
+
+void Intro::setDataDirectory(const QString &dataDir)
+{
+    ui->dataDirectory->setText(dataDir);
+    if(dataDir == GUIUtil::getDefaultDataDirectory())
+    {
+        ui->dataDirDefault->setChecked(true);
+        ui->dataDirectory->setEnabled(false);
+        ui->ellipsisButton->setEnabled(false);
+    } else {
+        ui->dataDirCustom->setChecked(true);
+        ui->dataDirectory->setEnabled(true);
+        ui->ellipsisButton->setEnabled(true);
+    }
+}
+
+bool Intro::showIfNeeded(bool& did_show_intro)
+{
+    did_show_intro = false;
+
+    QSettings settings;
+    /* If data directory provided on command line, no need to look at settings
+       or show a picking dialog */
+    if(!GetArg("-datadir", "").empty())
+        return true;
+    /* 1) Default data directory for operating system */
+    QString dataDir = GUIUtil::getDefaultDataDirectory();
+    /* 2) Allow QSettings to override default dir */
+    dataDir = settings.value("dataDir", dataDir).toString();
+
+    if(!fs::exists(GUIUtil::qstringToBoostPath(dataDir))
+            || GetBoolArg("-choosedatadir", DEFAULT_CHOOSE_DATADIR)
+            || settings.value("fReset", false).toBool()
+            || GetBoolArg("-resetguisettings", false))
+    {
+        /* Use selectParams here to guarantee Params() can be used by node interface when we implement it from
+           Bitcoin. */
+        try {
+            SelectParams(mapArgs.count("-testnet") ? CBaseChainParams::TESTNET : CBaseChainParams::MAIN);
+        } catch (const std::exception&) {
+            return false;
+        }
+
+        /* If current default data directory does not exist, let the user choose one */
+        Intro intro(0);
```

Comment:
We should be able to keep the free space heuristic from Bitcoin even though we don't have pruning since we maintain the estimates now (as of #1878).

```suggestion
        Intro intro(0, Params().AssumedBlockchainSize());
```

Reply:
Agreed.

Reply:
I just went and looked. Where are the estimates?

Reply:
https://github.com/gridcoin-community/Gridcoin-Research/blob/91a4862374e239165db54ef9c9865b902263b855/src/chainparams.cpp#L77

Reply:
I see m_assumed_blockchain_size in chainparams.h, but I don't see anything filling that out.

Reply:
Hmm... where is Bitcoin filling out blockchain_size_gb...

Reply:
Geez... it is just as simple as 2 GB for testnet and 4 GB for mainnet?

Reply:
That is a pretty bad heuristic.

Reply:
Ok. Put that back in.